x-pack/plugins/alerting/server/rules_client/lib/retry_if_bulk_conflicts.ts

```diff
@@ -16,50 +16,36 @@ import { RawRule } from '../../types';
 
 const MAX_RULES_IDS_IN_RETRY = 1000;
 
-export type BulkEnableOperation = (filter: KueryNode | null) => Promise<{
+interface BulkOperationReturn {
   errors: BulkOperationError[];
   rules: Array<SavedObjectsBulkUpdateObject<RawRule>>;
-  taskIdsToEnable: string[];
-}>;
-
-interface ReturnRetry {
-  errors: BulkOperationError[];
-  rules: Array<SavedObjectsBulkUpdateObject<RawRule>>;
-  taskIdsToEnable: string[];
+  accList: string[][];
 }
 
-/**
- * Retries BulkEnable requests
- * If in response are presents conflicted savedObjects(409 statusCode), this util constructs filter with failed SO ids and retries bulkEnable operation until
- * all SO updated or number of retries exceeded
- * @param logger
- * @param bulkEnableOperation
- * @param filter - KueryNode filter
- * @param retries - number of retries left
- * @param accErrors - accumulated conflict errors
- * @param accRules - accumulated enabled rules
- * @param accTaskIdsToEnable - accumulated task ids
- * @returns Promise<ReturnRetry>
- */
-
-export const retryIfBulkEnableConflicts = async (
+export const retryIfBulkConflicts = async (
+  action: 'DELETE' | 'ENABLE' | 'DISABLE',
   logger: Logger,
-  bulkEnableOperation: BulkEnableOperation,
+  bulkOperation: (filter: KueryNode | null) => Promise<BulkOperationReturn>,
   filter: KueryNode | null,
-  retries: number = RETRY_IF_CONFLICTS_ATTEMPTS,
+  accList?: string[][], // this list can include several accumulators, depends on the type of bulk operation
```

Comment:
I looked at the existing code, it doesn't seem like we're ever using the `accErrors` and `accRules` parameters (line 2655 in rules client) at all. From what I can tell, all we're doing is merging these results with the bulk operation results. Why can't we do that out of this function? IMO we should aim for helper functions to be concise and do fewer things.

So I'm not sure why we need to pass around a 2d array for retrying bulk operations. Let's try to make the function APIs clearer and more concise, especially when this is meant to be reused.

Also, even if we do decide to keep these, let's give these parameters better names, `accList` is not particularly descriptive.

Reply:
Ok I see we're doing this because we're calling it recursively, so I'm wondering if we should have an outer function (that 1 is public) that calls into the inner function, the inner function can call it self recursively with the params?

Reply:
Not sure I get why it will be better to have 2 functions instead of one.

Reply:
@JiaweiWu I've renamed accList here: defa02ad4cf513b1c4dc9bca71e0bf62ac50f0c4. It's the best I could come with :-) Maybe you have a better name.

Reply:
> Not sure I get why it will be better to have 2 functions instead of one.

Correct me if I'm wrong but we don't need to expose the `accList` variables to the person calling the function, since it exists to accumulate results of the recursive calls. Having 2 functions should make shorten the number of params that the caller needs to care about. :slightly_smiling_face:

Reply:
@JiaweiWu Does it look better for you?  05ea1a523eaa5e89d0014779abe4cbd98dd79eeb